server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistTestsStep.java

```diff
@@ -0,0 +1,214 @@
+    BatchReport.Component component = reportReader.readComponent(componentRef);
+    if (component.getIsTest() && reportReader.readTests(componentRef) != null) {
+      persistTestResults(component, context);
+    }
+
+    for (Integer childRef : component.getChildRefList()) {
+      recursivelyProcessComponent(context, childRef);
+    }
+  }
+
+  private void persistTestResults(BatchReport.Component component, TestContext context) {
+    ListMultimap<String, FileSourceDb.Test.CoveredFile> coveredFilesByName = loadCoverageDetails(component.getRef(), context);
+    List<FileSourceDb.Test> tests = buildDbTests(component, context, coveredFilesByName);
+
+    FileSourceDto existingDto = context.existingFileSourcesByUuid.get(component.getUuid());
+    long now = system.now();
+    if (existingDto != null) {
+      // update
+      existingDto
+        .setTestData(tests)
+        .setUpdatedAt(now);
+      dbClient.fileSourceDao().update(context.session, existingDto);
+    } else {
+      // insert
+      FileSourceDto newDto = new FileSourceDto()
+        .setTestData(tests)
+        .setFileUuid(component.getUuid())
+        .setProjectUuid(context.context.getProject().uuid())
+        .setDataType(Type.TEST)
+        .setCreatedAt(now)
+        .setUpdatedAt(now);
+      dbClient.fileSourceDao().insert(context.session, newDto);
+    }
+  }
+
+  private List<FileSourceDb.Test> buildDbTests(BatchReport.Component component, TestContext context, ListMultimap<String, FileSourceDb.Test.CoveredFile> coveredFilesByName) {
+    List<FileSourceDb.Test> tests = new ArrayList<>();
+    ReportIterator<BatchReport.Test> testIterator = new ReportIterator<>(context.reader.readTests(component.getRef()), BatchReport.Test.PARSER);
+    while (testIterator.hasNext()) {
+      BatchReport.Test batchTest = testIterator.next();
+      FileSourceDb.Test.Builder dbTest = FileSourceDb.Test.newBuilder();
+      dbTest.setType(batchTest.getType());
+      dbTest.setName(batchTest.getName());
+      if (batchTest.hasStacktrace()) {
+        dbTest.setStacktrace(batchTest.getStacktrace());
+      }
+      if (batchTest.hasStatus()) {
+        dbTest.setStatus(batchTest.getStatus());
+      }
+      if (batchTest.hasMsg()) {
+        dbTest.setMsg(batchTest.getMsg());
+      }
+      if (batchTest.hasExecutionTimeMs()) {
+        dbTest.setExecutionTimeMs(batchTest.getExecutionTimeMs());
+      }
+      List<FileSourceDb.Test.CoveredFile> coveredFiles = coveredFilesByName == null ? null : coveredFilesByName.get(batchTest.getName());
+      if (coveredFiles != null) {
+        dbTest.addAllCoveredFile(coveredFiles);
+      }
+      tests.add(dbTest.build());
+    }
+
+    return tests;
+  }
+
+  @CheckForNull
+  private ListMultimap<String, FileSourceDb.Test.CoveredFile> loadCoverageDetails(int testFileRef, TestContext context) {
+    File coverageDetailsFile = context.reader.readCoverageDetails(testFileRef);
+    if (coverageDetailsFile == null) {
+      return null;
```

Comment:
I'm not a bug fan of returning null for something that is returning a list. You should either return an empty list, or move the read of the report in persistTestResults.